source/containers.tex

```diff
@@ -14292,6 +14384,1710 @@
+                  less<typename KeyContainer::value_type>, KeyContainer, MappedContainer>;
+
+  template<class KeyContainer, class MappedContainer, class Allocator>
+    flat_map(KeyContainer, MappedContainer, Allocator)
+      -> flat_map<typename KeyContainer::value_type, typename MappedContainer::value_type,
+                  less<typename KeyContainer::value_type>, KeyContainer, MappedContainer>;
+
+  template<class KeyContainer, class MappedContainer>
+    flat_map(sorted_unique_t, KeyContainer, MappedContainer)
+      -> flat_map<typename KeyContainer::value_type, typename MappedContainer::value_type,
+                  less<typename KeyContainer::value_type>, KeyContainer, MappedContainer>;
+
+  template<class KeyContainer, class MappedContainer, class Allocator>
+    flat_map(sorted_unique_t, KeyContainer, MappedContainer, Allocator)
+      -> flat_map<typename KeyContainer::value_type, typename MappedContainer::value_type,
+                  less<typename KeyContainer::value_type>, KeyContainer, MappedContainer>;
+
+  template<class InputIterator, class Compare = less<@\exposid{iter-key-type}@<InputIterator>>>
+    flat_map(InputIterator, InputIterator, Compare = Compare())
+      -> flat_map<@\exposid{iter-key-type}@<InputIterator>, @\exposid{iter-mapped-type}@<InputIterator>, Compare>;
+
+  template<class InputIterator, class Compare = less<@\exposid{iter-key-type}@<InputIterator>>>
+    flat_map(sorted_unique_t, InputIterator, InputIterator, Compare = Compare())
+      -> flat_map<@\exposid{iter-key-type}@<InputIterator>, @\exposid{iter-mapped-type}@<InputIterator>, Compare>;
+
+  template<ranges::input_range R, class Compare = less<@\exposid{range-key-type}@<R>>,
+           class Allocator>
+    flat_map(from_range_t, R&&, Compare = Compare(), Allocator = Allocator())
+      -> flat_map<@\exposid{range-key-type}@<R>, @\exposid{range-mapped-type}@<R>, Compare>;
+
+  template<ranges::input_range R, class Allocator>
+    flat_map(from_range_t, R&&, Allocator)
+      -> flat_map<@\exposid{range-key-type}@<R>, @\exposid{range-mapped-type}@<R>>;
+
+  template<class Key, class T, class Compare = less<Key>>
+    flat_map(initializer_list<pair<Key, T>>, Compare = Compare())
+      -> flat_map<Key, T, Compare>;
+
+  template<class Key, class T, class Compare = less<Key>>
+    flat_map(sorted_unique_t, initializer_list<pair<Key, T>>, Compare = Compare())
+        -> flat_map<Key, T, Compare>;
+
+  template<class Key, class T, class Compare, class KeyContainer, class MappedContainer,
+            class Allocator>
+    struct uses_allocator<flat_map<Key, T, Compare, KeyContainer, MappedContainer>, Allocator>
+      : bool_constant<uses_allocator_v<KeyContainer, Allocator> &&
+                      uses_allocator_v<MappedContainer, Allocator>> { };
+}
+\end{codeblock}
+
+\pnum
+The member type \tcode{containers} has the data members and special members
+specified above.
+It has no base classes or members other than those specified.
+
```

Comment:
```suggestion

\pnum
This subclause makes use of the following exposition-only member functions of \tcode{flat_map}:
\begin{codeblock}
void @\exposid{erase-duplicate-elements}@() {  // expos
  auto zv = ranges::zip_view(c.keys, c.values);
  auto it = ranges::unique(zv, key_equiv(compare)).begin();
  auto dist = distance(zv.begin(), it);
  c.keys.erase(c.keys.begin() + dist, c.keys.end());
  c.values.erase(c.values.begin() + dist, c.values.end());
}

void @\exposid{append-elements}@(const auto& first, const auto& last) {  // expos
  for (; first != last; ++first) {
    value_type value = *first;
    c.keys.insert(c.keys.end(), std::move(value.first));
    c.values.insert(c.values.end(), std::move(value.second));
  }
}
\end{codeblock}

```

Reply:
It's a good suggestion, but I don't think it should be done as part of applying the paper to the draft. It's not what was approved, and can be changed later.

Reply:
Please send a pull request if you like to propose this change.